### ISSUE TITLE
Crash when opening h5py.File in context manager

### ISSUE THREAD

**Issue by andyfaff** [h5py/h5py#1010]
I am getting a hard interpreter crash when  `h5py.File` is used in a context manager, and I think it's to do with the way that the path is specified:

```
>>> import os.path
>>> import h5py

>>> # z is a network mount.
>>> pth = os.path.join('z:/current', 'myfile.h5')
>>> os.path.isfile(pth)
True

>>> # this is fine
>>> with h5py.File(pth, 'r') as f:
...    print(f)

>>> # Note - the slash is missing after 'z:'
>>> pth = os.path.join('z:current', 'myfile.h5')
>>> os.path.isfile(pth)
True

>>> # can open the file normally
>>> f = open(pth, 'rb')
>>> f.close

>>> # but this immediately crashes the interpreter
>>> with h5py.File(pth, 'r') as f:
...    print(f)
```

Summary of the h5py configuration
---------------------------------
h5py    2.7.1
HDF5    1.10.1
Python  3.6.2 |Continuum Analytics, Inc.| (default, Jul 20 2017, 12:30:02) [MSC v.1900 64 bit (AMD64)]
sys.platform    win32
sys.maxsize     9223372036854775807
numpy   1.14.1

**Comment by andyfaff** [comment]
Hmm, been a while since I wrote a C program. Never used hdf5 from C before.

**Comment by tacaswell** [comment]
Should we add some path handling to upper case drive letters automatically?

**Comment by vasole** [comment]
In the old days Andrew Collette was reluctant to change the behavior respect to the underlying HDF5 library. I do not know if that is still the case.

In that sense, if it is an HDF5 bug,  it would be enough to raise an exception/warning in order to

- avoid the crash of the interpreter 
- inform the user about what to do

I guess whatever you decide it will be fine.

**Comment by andyfaff** [comment]
> In that sense, if it is an HDF5 bug, it would be enough to raise an
exception/warning

A good idea. Is there a reason that the HDF5 library has to own the file
handle, rather than the h5py library passing a reference to it?

A.

**Comment by aragilar** [comment]
@andyfaff HDF5 controls the whole file process, we just get to pass it a name (unfortunately). This is a limitation of the HDF5 library.

I wonder if the best solution is for us to carry patches for HDF5 (we already need to do some setup to build on Windows anyway, and since I suspect most people use a precomplied h5py, we could just patch HDF5, both for this, and for using unicode paths). We should push that off until after 2.8 gets released though.

**Comment by tacaswell** [comment]
Sorry, finishing the 2.8 release keeps not making it high enough up my to-do list any given day....I'm feeling guilty about it if that makes anyone feel better.

It looks like this was fixed in the latest patch release (ftp://gamma.hdfgroup.org/pub/outgoing/hdf5/snapshots/v110/hdf5-1.10.2-snap10-RELEASE.txt this was HDFFV-10307).

I think the best fix is to just re-build the wheels and get 2.8 out.

**Comment by aparamon** [comment]
No more reproducible on h5py 2.9
